Mathlib/Topology/LocalHomeomorph.lean

```diff
@@ -0,0 +1,1420 @@
+#align local_homeomorph.nhds_within_target_inter LocalHomeomorph.nhdsWithin_target_inter
+
+theorem image_eq_target_inter_inv_preimage {s : Set α} (h : s ⊆ e.source) :
+    e '' s = e.target ∩ e.symm ⁻¹' s :=
+  e.toLocalEquiv.image_eq_target_inter_inv_preimage h
+#align local_homeomorph.image_eq_target_inter_inv_preimage LocalHomeomorph.image_eq_target_inter_inv_preimage
+
+theorem image_source_inter_eq' (s : Set α) : e '' (e.source ∩ s) = e.target ∩ e.symm ⁻¹' s :=
+  e.toLocalEquiv.image_source_inter_eq' s
+#align local_homeomorph.image_source_inter_eq' LocalHomeomorph.image_source_inter_eq'
+
+theorem image_source_inter_eq (s : Set α) :
+    e '' (e.source ∩ s) = e.target ∩ e.symm ⁻¹' (e.source ∩ s) :=
+  e.toLocalEquiv.image_source_inter_eq s
+#align local_homeomorph.image_source_inter_eq LocalHomeomorph.image_source_inter_eq
+
+theorem symm_image_eq_source_inter_preimage {s : Set β} (h : s ⊆ e.target) :
+    e.symm '' s = e.source ∩ e ⁻¹' s :=
+  e.symm.image_eq_target_inter_inv_preimage h
+#align local_homeomorph.symm_image_eq_source_inter_preimage LocalHomeomorph.symm_image_eq_source_inter_preimage
+
+theorem symm_image_target_inter_eq (s : Set β) :
+    e.symm '' (e.target ∩ s) = e.source ∩ e ⁻¹' (e.target ∩ s) :=
+  e.symm.image_source_inter_eq _
+#align local_homeomorph.symm_image_target_inter_eq LocalHomeomorph.symm_image_target_inter_eq
+
+theorem source_inter_preimage_inv_preimage (s : Set α) :
+    e.source ∩ e ⁻¹' (e.symm ⁻¹' s) = e.source ∩ s :=
+  e.toLocalEquiv.source_inter_preimage_inv_preimage s
+#align local_homeomorph.source_inter_preimage_inv_preimage LocalHomeomorph.source_inter_preimage_inv_preimage
+
+theorem target_inter_inv_preimage_preimage (s : Set β) :
+    e.target ∩ e.symm ⁻¹' (e ⁻¹' s) = e.target ∩ s :=
+  e.symm.source_inter_preimage_inv_preimage _
+#align local_homeomorph.target_inter_inv_preimage_preimage LocalHomeomorph.target_inter_inv_preimage_preimage
+
+theorem source_inter_preimage_target_inter (s : Set β) :
+    e.source ∩ e ⁻¹' (e.target ∩ s) = e.source ∩ e ⁻¹' s :=
+  e.toLocalEquiv.source_inter_preimage_target_inter s
+#align local_homeomorph.source_inter_preimage_target_inter LocalHomeomorph.source_inter_preimage_target_inter
+
+theorem image_source_eq_target (e : LocalHomeomorph α β) : e '' e.source = e.target :=
+  e.toLocalEquiv.image_source_eq_target
+#align local_homeomorph.image_source_eq_target LocalHomeomorph.image_source_eq_target
+
+theorem symm_image_target_eq_source (e : LocalHomeomorph α β) : e.symm '' e.target = e.source :=
+  e.symm.image_source_eq_target
+#align local_homeomorph.symm_image_target_eq_source LocalHomeomorph.symm_image_target_eq_source
+
+/-- Two local homeomorphisms are equal when they have equal `toFun`, `inv_fun` and `source`.
+It is not sufficient to have equal `toFun` and `source`, as this only determines `inv_fun` on
+the target. This would only be true for a weaker notion of equality, arguably the right one,
+called `eq_on_source`. -/
```

Comment:
```suggestion
/-- Two local homeomorphisms are equal when they have equal `toFun`, `invFun` and `source`.
It is not sufficient to have equal `toFun` and `source`, as this only determines `invFun` on
the target. This would only be true for a weaker notion of equality, arguably the right one,
called `eq_on_source`. -/
```